### PR TITLE
Feature: Encapsulated dynamic imports with error boundary and suspense

### LINKED ISSUE #18912: Frontend: Undefined behavior when loading old bundles after Grafana update
Potential scenario that will become more likely since we do more aggressive code splitting now (#18544):

When the browser loads Grafana, it keeps a manifest of bundles it can lazy load. Those bundles have fingerprints in the filename, and are version dependent. If now, while Grafana is loaded in browser, the grafana server is being updated, the old bundles are no longer available, i.e., the files in the manifest can no longer be loaded. This can result in unstable behavior. 

We should put a mechanism in place to handle failed bundle loads (due to upgrade). Other web apps do this too (e.g. by showing a top level message, "XYZ has been updated, please refresh the browser").

### PR DESCRIPTION
**Pull request by hugohaggmark**
**What this PR does / why we need it**:
This PR is an initial proposal to encapsulate dynamic imports with a common framework to make it more user friendly when Grafana fails to load chunks.

**Which issue(s) this PR fixes**:
Closes: #18912

**Special notes for your reviewer**:
Known limitation: This doesn't handle the plugin loader as that already throws an error that is displayed for user. Could be more user-friendly though.

To test ERROR case:
1. Login
2. Goto Datasources for instance `http://localhost:3000/datasources`
3. Add a comment in `ApiKeysPage`for instance and wait for compile to finish
4. Navigate to `http://localhost:3000/org/apikeys`
5. The following message should appear
image
6. Refresh page
7. `ApiKeysPage` should appear

To test the OK case you need slow down the network
image

You should be able to see the Suspense component while navigating around the application
image

### CODE REVIEW

File: public/app/routes/routes.ts [hunk 1]
```diff
@@ -77,7 +76,9 @@ export function setupAngularRoutes($routeProvider: route.IRouteProvider, $locati
       reloadOnSearch: false,
       resolve: {
         component: () =>
-          import(/* webpackChunkName: "SoloPanelPage" */ '../features/dashboard/containers/SoloPanelPage'),
+          SafeDynamicImport(
```
Comment (dprokop) [inline public/app/routes/routes.ts]: Cool stuff @hugohaggmark. Would this work if `SafeDynamicImport` accepted string instead? I assume Webpack would have problems understanding the magic comments.

**Comment by torkelo** [comment]
Great! Think we need to make the wording a bit more layman-friendly, not sure end users are going to understand what not being able to load a chunk means :)  A special view with a big reloads button for example maybe? 

Then we need something for loading data source plugins but in that case I think we need a modal will happen as you work / use a dashboard.

File: public/app/routes/routes.ts (hunk 1 above)
Comment (hugohaggmark) [reply public/app/routes/routes.ts]: Not sure how those magic comment stuff works in WebPack, could be worth looking at in future PR.

**Comment by hugohaggmark** [comment]
> Great! Think we need to make the wording a bit more layman-friendly, not sure end users are going to understand what not being able to load a chunk means :) A special view with a big reloads button for example maybe?
> 
> Then we need something for loading data source plugins but in that case I think we need a modal will happen as you work / use a dashboard.

Great points! Will see what I can do

**Comment by davkal** [comment]
Initially I was hoping for more of a subtle overlay that would keep the existing page (I've seen this in Google products), but that is too complex, then this works too.

Error message suggestion: "Grafana has been updated. Please refresh in your browser."
Also like the idea of a big "Refresh" call to action.

**Comment by torkelo** [comment]
> Initially I was hoping for more of a subtle overlay that would keep the existing page 

The problem is that it has not been loaded, and does not exist on the server anymore. ideally we could keep old js files around somehow after an upgrade, but without a cdn not sure how would do that.

File: public/app/routes/routes.ts (hunk 1 above)
Comment (kaydelaney) [reply public/app/routes/routes.ts]: Without something like macros I don't think it'd be possible, but maybe it's worth looking at https://github.com/kentcdodds/babel-plugin-macros or https://github.com/LeDDGroup/typescript-transform-macros ?

File: public/app/core/components/SafeDynamicImport.tsx [hunk 2]
```diff
@@ -1,20 +1,48 @@
...
+import { LoadingPlaceholder, ErrorBoundary, Button } from '@grafana/ui';
 
 export const LoadingChunkPlaceHolder: FunctionComponent = () => (
   <div className={cx('preloader')}>
-    <LoadingPlaceholder text={'Loading chunk...'} />
+    <LoadingPlaceholder text={'Loading...'} />
   </div>
 );
 
+function getAlertPageStyle() {
```
Comment (kaydelaney) [inline public/app/core/components/SafeDynamicImport.tsx]: Is there an advantage in having this as a function, versus just a variable?

Comment (hugohaggmark) [reply public/app/core/components/SafeDynamicImport.tsx]: Unfortunately just cut & paste in this case because I was in a hurry.

**Comment by torkelo** [comment]
Updated the PageLoader (also loading indicator for loading page content) to align them in looks. The placement is still different as the PageLoader is shown under page nav header (tabs). But can live with that :) 

The wording is tricky. After testing a bunch I ended with this: 
Screen Shot 2019-09-17 at 08 08 38

I switched to use word reload instead of refresh as that is what most browsers use I think (Chrome at least).

**Review by davkal — APPROVED** [review]
LGTM

**Pull request merged**